### ISSUE TITLE
Cross-compilation of lwt-2.4.5

### ISSUE THREAD

**Issue by mryau** [ocsigen/lwt#341]
I'm trying to adopt https://github.com/vouillon/opam-android-repository/tree/master/packages/android-lwt.2.4.5 from NDK cross-compiling suite to Yocto arm-oe-linux-gnueabi toolchain.
I see the following error:
```
- + /mnt/opam/4.02.1/bin/arm-oe-linux-gnueabi-ocamlfind ocamlc -ccopt -I/usr/include -ccopt -I/usr/include -ccopt -Isrc/unix -c src/unix/lwt_unix_stubs.c
- cc1: warning: include location "/usr/include" is unsafe for cross-compilation [-Wpoison-system-directories]
- In file included from /usr/include/netdb.h:27:0,
-                  from src/unix/lwt_unix_stubs.c:70:
- /usr/include/netinet/in.h:536:21: error: field 'ipi6_addr' has incomplete type
-      struct in6_addr ipi6_addr; /* src/dst IPv6 address */
-                      ^~~~~~~~~
- /usr/include/netinet/in.h:543:25: error: field 'ip6m_addr' has incomplete type
-      struct sockaddr_in6 ip6m_addr; /* dst address including zone ID */
-                          ^~~~~~~~~
- In file included from src/unix/lwt_unix_stubs.c:97:0:
- src/unix/lwt_unix_unix.c:192:23: error: field 's_inet6' has incomplete type
-    struct sockaddr_in6 s_inet6;
-                        ^~~~~~~
- In file included from src/unix/lwt_unix_stubs.c:97:0:
- src/unix/lwt_unix_unix.c: In function 'alloc_one_addr6':
- src/unix/lwt_unix_unix.c:1727:19: error: storage size of 'addr' isn't known
-    struct in6_addr addr;
-                    ^~~~
- src/unix/lwt_unix_unix.c:1727:19: warning: unused variable 'addr' [-Wunused-variable]
- src/unix/lwt_unix_unix.c:1730:1: warning: control reaches end of non-void function [-Wreturn-type]
-  }
-  ^
- Command exited with code 2.
```
Setting environment variables CC,LD and CPP to point to correct programs with "--sysroot=..." has ho effect for setup.ml.
How can I set something like "--sysroot=..."option for gcc in setup.ml command line or, maybe in ocamlfind arguments?

**Comment by mfp** [comment]
On Tue, Apr 18, 2017 at 05:16:21AM -0700, mryau wrote:
> I'm trying to adopt https://github.com/vouillon/opam-android-repository/tree/master/packages/android-lwt.2.4.5 from NDK cross-compiling suite to Yocto arm-oe-linux-gnueabi toolchain.
[...]
> Setting environment variables CC,LD and CPP to point to correct programs with "--sysroot=..." has ho effect for setup.ml.
> How can I set something like "--sysroot=..."option for gcc in setup.ml command line or, maybe in ocamlfind arguments?

Can you try to override bytecomp_c_compiler and native_c_compiler with

   ocaml setup.ml configure \
      --override bytecomp_c_compiler "gcc .... --sysroot=..." \
      --override native_c_compiler "gcc .... --sysroot=..."

?

The current values (that include the required gcc options) are in setup.data,
generated by setup.ml's configure phase.

Alternatively, you could try to pass -ccopt "--sysroot=..."  and -cclib ...
to ocamlopt/ocamlc externally via OCAMLPARAM.

-- 
Mauricio Fernández

**Comment by mryau** [comment]
Thank you for fast response.
I modified my opam file aoordinging to your reccomendation:
```
build: [
  ["oasis" "setup"]
  [
    "ocaml"
    "setup.ml"
    "-debug"
    "-configure"
    "--override"
    "bytecomp_c_compiler"
    "arm-oe-linux-gnueabi-gcc  -march=armv7-a -mfpu=vfpv3  -mfloat-abi=softfp --sysroot=/home/ssp/build/tmp-glibc/sysroots/wolverine_v1 -fno-defer-pop -Wall -D_FILE_OFFSET_BITS=64 -D_REENTRANT -fPIC"
    "--override"
    "native_c_compiler"
    "arm-oe-linux-gnueabi-gcc  -march=armv7-a -mfpu=vfpv3  -mfloat-abi=softfp --sysroot=/home/ssp/build/tmp-glibc/sysroots/wolverine_v1 -Wall -D_FILE_OFFSET_BITS=64 -D_REENTRANT"
    "--override"
    "ocamlfind"
    "%{bin}%/arm-oe-linux-gnueabi-ocamlfind"
    "--enable-ssl"
  ]
  ["ocaml" "setup.ml" "-build"]
  ["ocaml" "setup.ml" "-install"]
]
```
and setted up OCAMLPARAM variable:
export OCAMLPARAM="_,ccopts=--sysroot=/home/ssp/build/tmp-glibc/sysroots/wolverine_v1,cclib=--sysroot=/home/ssp/build/tmp-glibc/sysroots/wolverine_v1"
but `ocaml setup.ml -build` says same error:
```
+ /mnt/opam/4.02.1/bin/arm-oe-linux-gnueabi-ocamlfind ocamlc -ccopt -I/usr/include -ccopt -I/usr/include -ccopt -Isrc/unix -c src/unix/lwt_unix_stubs.c
cc1: warning: include location "/usr/include" is unsafe for cross-compilation [-Wpoison-system-directories]
In file included from /usr/include/netdb.h:27:0,
                 from src/unix/lwt_unix_stubs.c:70:
/usr/include/netinet/in.h:536:21: error: field 'ipi6_addr' has incomplete type
     struct in6_addr ipi6_addr; /* src/dst IPv6 address */
                     ^~~~~~~~~
/usr/include/netinet/in.h:543:25: error: field 'ip6m_addr' has incomplete type
     struct sockaddr_in6 ip6m_addr; /* dst address including zone ID */
                         ^~~~~~~~~
In file included from src/unix/lwt_unix_stubs.c:97:0:
src/unix/lwt_unix_unix.c:192:23: error: field 's_inet6' has incomplete type
   struct sockaddr_in6 s_inet6;
                       ^~~~~~~
In file included from src/unix/lwt_unix_stubs.c:97:0:
src/unix/lwt_unix_unix.c: In function 'alloc_one_addr6':
src/unix/lwt_unix_unix.c:1727:19: error: storage size of 'addr' isn't known
   struct in6_addr addr;
                   ^~~~
src/unix/lwt_unix_unix.c:1727:19: warning: unused variable 'addr' [-Wunused-variable]
src/unix/lwt_unix_unix.c:1730:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
Command exited with code 2.
Compilation unsuccessful after building 89 targets (85 cached) in 00:00:00.
E: Failure("Command ''/mnt/opam/4.02.1/bin/ocamlbuild' syntax/optcomp.cma syntax/optcomp.cmxa syntax/optcomp.a syntax/optcomp.cmxs src/core/lwt.cma src/core/lwt.cmxa src/core/lwt.a src/core/lwt.cmxs src/logger/lwt-log.cma src/logger/lwt-log.cmxa src/logger/lwt-log.a src/logger/lwt-log.cmxs src/unix/liblwt-unix_stubs.a src/unix/dlllwt-unix_stubs.so src/unix/lwt-unix.cma src/unix/lwt-unix.cmxa src/unix/lwt-unix.a src/unix/lwt-unix.cmxs src/simple_top/lwt-simple-top.cma src/simple_top/lwt-simple-top.cmxa src/simple_top/lwt-simple-top.a src/simple_top/lwt-simple-top.cmxs src/preemptive/lwt-preemptive.cma src/preemptive/lwt-preemptive.cmxa src/preemptive/lwt-preemptive.a src/preemptive/lwt-preemptive.cmxs src/extra/lwt-extra.cma src/extra/lwt-extra.cmxa src/extra/lwt-extra.a src/extra/lwt-extra.cmxs src/ssl/lwt-ssl.cma src/ssl/lwt-ssl.cmxa src/ssl/lwt-ssl.a src/ssl/lwt-ssl.cmxs syntax/lwt-syntax.cma syntax/lwt-syntax.cmxa syntax/lwt-syntax.a syntax/lwt-syntax.cmxs syntax/lwt-syntax-options.cma syntax/lwt-syntax-options.cmxa syntax/lwt-syntax-options.a syntax/lwt-syntax-options.cmxs syntax/lwt-syntax-log.cma syntax/lwt-syntax-log.cmxa syntax/lwt-syntax-log.a syntax/lwt-syntax-log.cmxs examples/unix/logging.native examples/unix/relay.native examples/unix/parallelize.native -tag debug' terminated with error code 10")
```

**Comment by aantron** [comment]
Hmm, changing `bytecomp_c_compiler` and `native_c_compiler` to an obviously-wrong value had no effect on the build, so I guess those won't work (and I verified that the wrong value ended up in `setup.data`).

You may be able to pass extra arguments by modifying `myocamlbuild.ml`. If going down that route, you need to add something like this in [`After_rules`](https://github.com/ocsigen/lwt/blob/2.4.5/myocamlbuild.ml#L361):

    flag ["compile"; "c"] (S [A "-ccopt"; A "foo"; A "-ccopt"; A "bar"])

Of course, replace `"foo"` and `"bar"` with what you actually need.

**Comment by mryau** [comment]
It seems, I've found problem code: at -configure phase setup.ml executes discover.ml with wrong -ocamlc parameter as following:
```
E: Failure("Command 'ocaml discover.ml -ocamlc /mnt/opam/4.02.1/bin/ocamlc.opt -ext-obj .o -exec-name a.out -use-libev true -os-type Unix -use-glib false -ccomp-type cc -use-pthread true -use-unix true -android-target false -libev_default true' terminated with error code 1")
...
testing for eventfd: ........................... unavailable
testing for fd passing: ........................ unavailable
testing for sched_getcpu: ...................... unavailable
testing for affinity getting/setting: .......... unavailable
testing for credentials getting (Linux): ....... unavailable
testing for credentials getting (NetBSD): ...... unavailable
testing for credentials getting (OpenBSD): ..... unavailable
testing for credentials getting (FreeBSD): ..... unavailable
testing for credentials getting (getpeereid): .. unavailable
testing for fdatasync: ......................... unavailable
testing for netdb_reentrant: ................... unavailable
...
```
So, it did not cofigured code in src/unix/lwt_unix_stubs.c to use
_GNU_SOURCE and HAVE_GET_CREDENTIALS_LINUX
Currently I simply simulate right results of test by returning true in code of discover.ml

**Comment by aantron** [comment]
@mryau that command line doesn't look like that in `master` anymore, it was changed in #109, which was released as part of 2.4.7. I don't know how practical it is for you to use a newer version of Lwt, but is it possible that #109 addresses your problem? I guess you could just patch your Lwt.

There also some Android packages [here](https://github.com/whitequark/opam-cross-android/tree/master/packages), including Lwt 2.6.0, but, again, I don't know if you have seen these already, or if they are relevant to what you are doing.

**Comment by aantron** [comment]
@mryau I'm closing this as "hopefully fixed by newer Lwt," but please do reopen if you have more information, etc.
